### PR TITLE
Use static_link_msvcrt (libcrt, e.g. /MT) for Windows build

### PR DESCRIPTION
**Pull request by achasveachas**
Description:
    Use static_link_msvcrt (libcrt, e.g. /MT) for Windows build
    
    This patch introduces necessary flag overrides to many foreign_cc
    components and toggles the compile flag (/MT) for these packages;
      cares, curl, libevent, nghttp2, yaml and zlib
    
    This has the side effect of having a working cares linkage for windows.
    To expedite development and test review, observe cached test results.

    This should have zero impact on Linux

Risk Level: Low
Testing: Local on Windows
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by junr03** [comment]
@lizan tagging you as you have been working on the windows build.

File: bazel/foreign_cc/nghttp2.patch [hunk 1]
```diff
@@ -15,15 +27,31 @@ index e70d2d49..f9c5340e 100644
...
-     VERSION ${LT_VERSION} SOVERSION ${LT_SOVERSION}
-     ARCHIVE_OUTPUT_NAME nghttp2
-+    ARCHIVE_OUTPUT_NAME nghttp2_static
+@@ -385,6 +389,15 @@ else()
    )
-   target_compile_definitions(nghttp2_static PUBLIC "-DNGHTTP2_STATICLIB")
-   if(ENABLE_STATIC_LIB)
+ endif()
+ 
++if(ENABLE_STATIC_CRT)
```
Comment (lizan) [inline bazel/foreign_cc/nghttp2.patch]: Can you open a PR in upstream too for this change?

File: bazel/grpc_grpc.patch [hunk 2]
```diff
@@ -0,0 +1,43 @@
+diff --git a/BUILD b/BUILD
```
Comment (lizan) [inline bazel/grpc_grpc.patch]: why do you need this patch? Isn't `CARES_STATICLIB` defined in the cc_library `ares` so it should propagate?

Comment (wrowe) [reply bazel/grpc_grpc.patch]: There's a small misunderstanding here.

Using the static libcrt.lib and avoiding the linkage to msvcrt.dll has nothing to do at all with whether a given library is a static (non-PIC) source compilation. This patch alters the Microsoft CRT we bind to, as opposed to the object files we are compiling.

File: bazel/foreign_cc/nghttp2.patch (hunk 1 above)
Comment (wrowe) [reply bazel/foreign_cc/nghttp2.patch]: I've opened three representing the all of the changes in nghttp2.patch (including one correction).

Correct ssize_t on Windows
https://github.com/nghttp2/nghttp2/pull/1395

Avoid filename collision of static and dynamic lib
https://github.com/nghttp2/nghttp2/pull/1394
(Dropping the erroneous ARCHIVE_OUTPUT_NAME nghttp2 value)

https://github.com/nghttp2/nghttp2/pull/1393
Add new flag ENABLE_STATIC_CRT for Windows

**Comment by achasveachas** [comment]
We fixed the merge conflict in `windows/.bazelrc`.

@lizan ready for another go-over.

**Comment by lizan** [comment]
please do not rebase and force push, change is messed up somehow, can you clean it up, or open another PR and close this one.

**Comment by wrowe** [comment]
Superseded by PR #8280 per @lizan's request. Apologies for our merge conflict resolution.

**Pull request closed without merging**